reactive/kubernetes_worker.py

```diff
@@ -1325,3 +1331,19 @@ def get_registry_location():
         registry = ""
 
     return registry
+
+
+@when('coredns.ready', 'kube-control.dns.available')
+def set_cluster_dns():
+    """Set ClusterDNS in the kubelet config.
+
+    If a relation with the CoreDNS k8s charm is added its service
+    IP will be added in the kubelet config as the ClusterDNS argument
+    and the kubelet will be restarted.
+    """
+    kube_control = endpoint_from_flag('kube-control.dns.available')
```

Comment:
I'm confused by the fact that we seem to still be relying on dns info from the kube-control relation here. I'm not sure that's correct. Does the code still work if 'dns-provider' == 'none' in the master charm config?

Reply:
Yes it still works with 'dns-provider' == 'none'. I mostly used it because I wanted to make sure I wasn't missing any value in the dns object that i get from `kube_control.get_dns()`. If it has just the two values that I set (`sdn-ip`,`enable-kube-dns`) I can create it on my own. 

Also, the ClusterDNS variable in the kubelet config is a list, so it could have more than one DNS server. Is that something that we should take in account?

Reply:
I don't think we need to support multiple dns servers right now - we can add that later if needed.

I do want to make sure that the cluster-dns relation doesn't have an implicit dependency on the kube-control relation. The cluster-dns relation should be able to stand on its own; in other words, cluster-dns shouldn't have to consult kube-control to get dns info.

@Cynerva check me on this please. ^ I'd appreciate a second opinion.

Reply:
The info that's coming from kube-control is the DNS domain, which is set in the kubernetes-master [dns_domain charm config](https://github.com/charmed-kubernetes/charm-kubernetes-master/blob/552aa105024d2f1e3ae31674285cd202f1a368d7/config.yaml#L80-L83).

kubernetes-master, kubernetes-worker, and coredns all need to use the same DNS domain:
* kubernetes-master has to include kubernetes.{dns_domain} and kubernetes.default.svc.{dns_domain} as Subject Alternative Names in TLS certs.
* kubernetes-worker has to configure kubelet with clusterDomain.
* coredns has to be configured with the CLUSTER_DOMAIN it's serving.

Having the worker use the DNS domain from kube-control ensures that kubernetes-master and kubernetes-worker are in agreement. I see no problem there.

But there is nothing to ensure that kubernetes-master and coredns will use the same domain. Are we missing a relation between kubernetes-master and coredns?

Reply:
Currently `dns_domain` is a configurable value in the coredns charm. Although another relation might be a better solution.